### PR TITLE
Fix football results page race condition

### PR DESCRIPTION
**Pull request by jfsoul**
## What does this change?
This change fixes a race condition between 2 different processes that refresh `competition` data in `sport`.  

We have a series of `competition` agents which store result, fixture, start date and other data for particular competitions (note: these agents do not support multiple seasons).  The data contained within these agents is refreshed periodically using PA endpoints like `/fixtures` `/results` etc.  Importantly, `startDate` is refreshed separately by another process, but is in turn used in the refresh of results.

This other process for refreshing competition `startDate` which polls the PA competition endpoint is run on a separate schedule and therefore we can't rely on a competition to have a start date at all when the results are refreshed.  This race condition has been accounted for all along by defaulting the start date to 30 days ago.

Unfortunately when PA release their new season start dates, we prioritise the most recent season start date, which means the results may be empty depending on the timing of the competition result refresh and start date refresh.

~This PR fixes that issue by only taking the most recent _started_ season for the result refresh.  This could have consequences for the fixtures, but it turns out that those don't use the start date at all.  The only other affected construct is `wallcharts` which are for competitions that should only ever have one season anyway e.g. World Cups (afaik - we don't have any atm to test).~

## What is the value of this and can you measure success?
Results page is consistent once more.

## Does this affect other platforms - Amp, Apps, etc?
No

## Tested in CODE?
I plan to.

### CODE REVIEW

File: sport/app/football/feed/Competitions.scala [hunk 1]
```diff
@@ -149,19 +152,19 @@ class CompetitionsService(val footballClient: FootballClient, competitionDefinit
 
   override def competitions: Seq[Competition] = competitionAgents.map(_.competition)
 
-  def refreshCompetitionAgent(id: String): Option[Unit] = competitionAgents
```
Comment (jfsoul) [inline sport/app/football/feed/Competitions.scala]: `Option[Unit]` 🙈

File: sport/app/football/feed/Competitions.scala [hunk 2]
```diff
@@ -149,19 +152,19 @@ class CompetitionsService(val footballClient: FootballClient, competitionDefinit
...
       log.info(s"Completed refresh of competition '${c.competition.fullName}': currently ${c.competition.matches.length} matches")
     }
 
   def refreshCompetitionData() = {
     log.info("Refreshing competition data")
     footballClient.competitions.map { allComps =>
-      mostRecentCompetitionSeasons(allComps).map { season =>
-        competitionAgents.find(_.competition.id == season.id).map { agent =>
-          val newCompetition = agent.competition.startDate match {
+      mostRecentCompetitionSeason(allComps).foreach { season =>
```
Comment (jfsoul) [inline sport/app/football/feed/Competitions.scala]: fixed some maps with `Unit` return type warnings

**Review by TBonnin — COMMENTED** [review]
Looks fine to me. 
Have you talked to @JustinPinner who is the last one touching this logic. 
I find the PA logic to be quite fragile sometimes, I am worried we have overlooked potential problems with this change

**Comment by JustinPinner** [comment]
This change looks reasonable to me. I had a chat with @jfsoul earlier about the previous version of this code, but I couldn't remember the scenario that required having the two most recent seasons for a competition beyond season boundaries. I'd suggest talking to the sports desk before committing to these changes, as they will probably be able to suggest a non-obvious situation where it could still be required.

File: sport/app/football/feed/Competitions.scala [hunk 3]
```diff
@@ -134,13 +134,14 @@ class CompetitionsService(val footballClient: FootballClient, competitionDefinit
...
   )
 
-  private def mostRecentCompetitionSeasons(competitions: List[Season]): List[Season] = {
-    competitionDefinitions.flatMap{ compDef =>
-      competitions.filter(_.competitionId == compDef.id)
+  private def mostRecentCompetitionSeason(competitions: List[Season]): List[Season] =
+    competitionDefinitions.flatMap { compDef =>
+      competitions
+        .filter(_.competitionId == compDef.id)
+        .filter(_.startDate.isBefore(LocalDate.now()))
```
Comment (TBonnin) [inline sport/app/football/feed/Competitions.scala]: Maybe worth a comment on this line to explain why we filter future season?

Comment (jfsoul) [inline sport/app/football/feed/Competitions.scala]: Yeah it needs a comment, but I am rethinking this fix a little bit first.

**Comment by jfsoul** [comment]
@TBonnin @JustinPinner I've updated this PR so that it's practically the same as before except we take the _oldest_ of the most recent 2 seasons for the start date from which to fetch results.

This is very similar to the behaviour I had with the previous fix, but once the new season starts we will still get match nav info for matches from the previous season (for example).

**Comment by TBonnin** [comment]
Assuming we are in the middle of one season, would this also load the results of last year?

**Comment by jfsoul** [comment]
Yes, but I think that is what we want for a few reasons e.g.

- At the start of a new season you can still page back to last season's results
- Match report navs for last season still work (if the match is older than that then the ajax request gets 404)

Arguably we should do more than 2 seasons, but I don't want to change that right now.

**Comment by TBonnin** [comment]
Ok. Fine with me

**Review by TBonnin — APPROVED** [review]

**Pull request merged**